### PR TITLE
Update Mobile UCR Versions for Domains (Script)

### PR DESCRIPTION
**Pull request by zandre-eng**
## Product Description

No user-facing changes.

## Technical Summary

Link to ticket [here](https://dimagi.atlassian.net/browse/SC-3914).

**Please note:** This PR should not be merged and is only created for review purposes. The script contained in this PR will be run as a once-off script.

This PR is the code for a once-off script that will be run to update the Mobile UCR versions of all domains from V1 to V2. This will only be done for domains that do not have any manual references to V1 fixtures. Domains that do contain V1 fixtures will be flagged and will need to be handled separately.

## Feature Flag

`MOBILE_UCR`

### Safety story

Local testing

### Automated test coverage

N/A

### QA Plan

No QA planned.

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/mobile_ucr_v2_update_script.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+
+skip_domains = []
+
+
+def process():
+    try:
+        processed_domains = read_ndjson_file(PROCESSED_DOMAINS_PATH)
+    except FileNotFoundError:
+        processed_domains = []
```
Comment (kaapstorm) [inline corehq/mobile_ucr_v2_update_script.py:77]: Nit: Better to use a set: If you're checking for membership of a group, a set will be O(1) but a list will be O(n). This applies to `skip_domains` too.

**Review by kaapstorm — APPROVED** [review]
Just three suggestions, nothing blocking. I am happy with your approach.

File: corehq/mobile_ucr_v2_update_script.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    app.save()
+
+
+PROCESSED_DOMAINS_PATH = '/home/zandre/cchq/updated_domains.ndjson'
+LOG_FILE = '/home/zandre/cchq/update_to_v2_ucr_script.log'
+
+V1_FIXTURE_IDENTIFIER = 'src="jr://fixture/commcare:reports'
+V1_FIXTURE_PATTERN = r'<.*src="jr://fixture/commcare:reports.*>'
+V1_REFERENCES_PATTERN = r"<.*instance\('reports'\)/reports/.*>"
+V1_ALL_REFERENCES = f"{V1_FIXTURE_PATTERN}|{V1_REFERENCES_PATTERN}"
```
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py]: This is useful to know, thanks! Addressed in [0392d0f](https://github.com/dimagi/commcare-hq/pull/35158/commits/0392d0f11fb111b9ed45e36d86082903b37fdd52).

File: corehq/mobile_ucr_v2_update_script.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    with open(path, 'r') as file:
+        return [json.loads(line) for line in file.readlines()]
+
+
+def has_non_v2_form(domain, app, log_file):
+    for form in app.get_forms():
+        save_in_log(log_file, f"Processing Form: {domain}: {form.name}")
+        # The second condition should always be False if the first one is
+        # but just as a precaution we check for it
+        if V1_FIXTURE_IDENTIFIER in form.source or re.findall(V1_ALL_REFERENCES, form.source):
```
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py]: Yeah this sounds like the better approach. Addressed in [0392d0f](https://github.com/dimagi/commcare-hq/pull/35158/commits/0392d0f11fb111b9ed45e36d86082903b37fdd52).

File: corehq/mobile_ucr_v2_update_script.py (hunk 1 above)
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py]: This is a good idea. Addressed in [b3e4252](https://github.com/dimagi/commcare-hq/pull/35158/commits/b3e42528e1db8f52a77b9298b8592f6493bd9426).

**Comment by gherceg** [comment]
Excited to see this happening! I see there are some tickets ([here](https://dimagi.atlassian.net/browse/SC-3937) and [here](https://dimagi.atlassian.net/browse/SC-3760)) already created to default to V2, but I can't tell if those changes have been deployed/what the current behavior on production is. It would make sense to me to ensure we default to V2 prior to running this, so just making sure that is the plan or if not, there is a reason why.

**Comment by zandre-eng** [comment]
Thanks for sharing the concern @gherceg! I definitely agree this would be important to do beforehand and can confirm that we do have a plan outlined to first ensure that domains aren't able to go back to V1 before running the migration script from this PR. This also includes the related code changes from [this](https://dimagi.atlassian.net/browse/SC-3915) Jira ticket to handle when users are reverting/publishing apps with V1 references.

**Review by kaapstorm — APPROVED** [review]

File: corehq/mobile_ucr_v2_update_script.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+def process():
+    try:
+        processed_domains = read_ndjson_file(PROCESSED_DOMAINS_PATH)
+    except FileNotFoundError:
+        processed_domains = set()
+
+    mobile_ucr_domains = find_domains_with_toggle_enabled(MOBILE_UCR)
+
+    log_file = open(LOG_FILE, 'a')
```
Comment (mkangia) [inline corehq/mobile_ucr_v2_update_script.py:81]: I believe this would leave the connection to the file open even after the script exits abruptly, so we should this with `with` option. This way you also don't need to worry about explicitly calling close.
`with open(LOG_FILE, 'a') as log_file:`

Does mode 'a' create the file if its missing?

File: corehq/mobile_ucr_v2_update_script.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        if domain in skip_domains:
+            save_in_log(log_file, f"Skipped domain: {domain}")
+            continue
+
+        save_in_log(log_file, f"Processing domain: {domain} ...")
+        app_ids = list(get_latest_app_ids_and_versions(domain))
+        apps = get_apps_by_id(domain, app_ids)
+        for app in apps:
+            try:
+                # Don't look at app.is_released since the latest version might not be released yet
```
Comment (mkangia) [inline corehq/mobile_ucr_v2_update_script.py:98]: Confirming that we are just picking app and not releases.
I believe one way to confirm that is by checking `copy_of` property on the application. Is it blank for applications and not blank for releases. it would be a good check to have though I assume you are anyway only picking up applications and not releases.

File: corehq/mobile_ucr_v2_update_script.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    app.save()
+
+
+PROCESSED_DOMAINS_PATH = '/home/zandre/cchq/updated_domains.ndjson'
+LOG_FILE = '/home/zandre/cchq/update_to_v2_ucr_script.log'
+
+V1_FIXTURE_IDENTIFIER = 'src="jr://fixture/commcare:reports'
+V1_FIXTURE_PATTERN = r'<.*src="jr://fixture/commcare:reports.*>'
+V1_REFERENCES_PATTERN = r"<.*instance\('reports'\)/reports/.*>"
+RE_V1_ALL_REFERENCES = re.compile(f"{V1_FIXTURE_PATTERN}|{V1_REFERENCES_PATTERN}")
```
Comment (mkangia) [inline corehq/mobile_ucr_v2_update_script.py:67]: I trust this has been tested so it matches as needed.

**Review by mkangia — APPROVED** [review]
Non blocking comments. 

Would be nice to have a -dry-run option to the script so it can be run without doing actual updates to see which apps it would update or not update eventually. I have always found that useful, specifically on production

File: corehq/mobile_ucr_v2_update_script.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+
+skip_domains = set()
+
+
+def process():
```
Comment (mkangia) [inline corehq/mobile_ucr_v2_update_script.py:73]: nit: the flow of code in the file made it a bit hard to follow along. Just reminding on keeping things in a file in logical order. I am big fan on [vertical formatting](https://www.baeldung.com/cs/clean-code-formatting#vertical-formatting)

File: corehq/mobile_ucr_v2_update_script.py (hunk 4 above)
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py]: Looking through the code, I can remove the need for passing the log file around to the various helper functions. Since we already know what the file path is, we can just use this directly in the helper function for adding to the log. Addressed in [27c1c25](https://github.com/dimagi/commcare-hq/pull/35158/commits/27c1c2501790b14dfbbe049e3dff4d4c68a88656).

> Does mode 'a' create the file if its missing?

It does. [Here](https://phoenixnap.com/kb/file-handling-in-python#:~:text=Append%20mode%20adds%20information%20to,not%20clear%20a%20file's%20contents.) is an article for more info on the append mode.

File: corehq/mobile_ucr_v2_update_script.py (hunk 5 above)
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py:63]: > Confirming that we are just picking app and not releases.

Could you please elaborate on what you mean by app vs release? Are you referring to app versions that have been marked as "Released" or are you referring to the app versions themselves?

This comment refers that we are picking the latest version of each app and we don't want to consider `app.is_released` since the latest version of an app might still be set to "Test" and not "Released"

File: corehq/mobile_ucr_v2_update_script.py (hunk 6 above)
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py:35]: These regex patterns come from [this](https://gist.github.com/ajeety4/aa3a0604f9597c8fc172b23475cbad57) script which has been used to analyze all the domains with V1 references. Based on the results of the script that was run, these regex patterns have successfully picked up domains with V1 references.

These references were taken from [this](https://dimagi.atlassian.net/wiki/spaces/GS/pages/2146604299/Moving+to+Mobile+UCR+V2) Confluence page, which shows what needs to change to migrate from V1->V2.

**Comment by zandre-eng** [comment]
> Would be nice to have a -dry-run option to the script so it can be run without doing actual updates to see which apps it would update or not update eventually.

@mkangia I think this is a great idea. I have added a dry run option in [20d9401](https://github.com/dimagi/commcare-hq/pull/35158/commits/20d94010be2eaec504de8e804ffd266d1abd35bb).

File: corehq/mobile_ucr_v2_update_script.py (hunk 5 above)
Comment (mkangia) [reply corehq/mobile_ucr_v2_update_script.py:63]: > Are you referring to app versions that have been marked as "Released" or are you referring to the app versions themselves?

app versions or app builds.

File: corehq/mobile_ucr_v2_update_script.py [hunk 8]
```diff
@@ -37,7 +37,7 @@
 skip_domains = set()
 
 
-def process():
+def process(dry_run=False):
```
Comment (mkangia) [inline corehq/mobile_ucr_v2_update_script.py:40]: nit: totally up to you and fine to leave it as it is for this one, I usually keep dry run as True by default, it avoids accidental updates for me, but it is also fine either way.

**Review by mkangia — APPROVED** [review]

File: corehq/mobile_ucr_v2_update_script.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        return set(json.loads(line) for line in file.readlines())
+
+
+def has_non_v2_form(domain, app):
+    for form in app.get_forms():
+        save_in_log(f"Processing Form: {domain}: {form.name}")
+        # The second condition should always be False if the first one is
+        # but just as a precaution we check for it
+        if V1_FIXTURE_IDENTIFIER in form.source or RE_V1_ALL_REFERENCES.search(form.source):
+            save_in_log(f"App Contains V1 Refs: {domain}: {app.name}")
```
Comment (ajeety4) [inline corehq/mobile_ucr_v2_update_script.py:101]: Do you think we should add the form name or form id in the log as well ? 
Based on how we plan for assisting projects to update their manual references, this might be helpful. It does have a downside of checking for all the forms thereby increasing the script execution time.

**Review by ajeety4 — APPROVED** [review]
Looks good to me. 
Just a non-blocking comment. 

Also I think it would be good to test that after updating the app to v2 (through script or manually), it does not break anything. Preferably one of the app from echis domain.

File: corehq/mobile_ucr_v2_update_script.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+)
+from corehq.apps.app_manager.const import MOBILE_UCR_VERSION_2
+from corehq.toggles import MOBILE_UCR
+from corehq.toggles.shortcuts import find_domains_with_toggle_enabled
+from corehq.util.log import with_progress_bar
+
+
+PROCESSED_DOMAINS_PATH = '/home/zandre/cchq/updated_domains.ndjson'
+LOG_FILE = '/home/zandre/cchq/update_to_v2_ucr_script.log'
+
```
Comment (ajeety4) [inline corehq/mobile_ucr_v2_update_script.py:31]: Hey @zandre-eng , based on the [comment here](https://github.com/dimagi/commcare-hq/pull/35168#discussion_r1800689494) in the other PR, requesting changes to cover the Case List/Detail scenario

```suggestion

# Pattern for references in case list/detail. 
# The below instance needs to be referred in a fixture setup and is then later used in Case List/Detail filters and columns.
V1_CASE_LIST_REFERENCES_PATTERN = r"instance\('commcare:reports'\)/reports/report\[@id='.*'\]/rows/row"
RE_V1_CASE_LIST_REFERENCES_PATTERN = re.compile(V1_CASE_LIST_REFERENCES_PATTERN)
```

File: corehq/mobile_ucr_v2_update_script.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        save_in_log(f"Processing domain: {domain} ...")
+        app_ids = list(get_latest_app_ids_and_versions(domain))
+        apps = get_apps_by_id(domain, app_ids)
+        for app in apps:
+            try:
+                # Don't look at app.is_released since the latest version might not be released yet
+                if app.mobile_ucr_restore_version != '2.0':
+                    save_in_log(f"Processing App: {domain}: {app.name}: {app.id}")
+                    if not has_non_v2_form(domain, app):
```
Comment (ajeety4) [inline corehq/mobile_ucr_v2_update_script.py:66]: Based on [this](https://github.com/dimagi/commcare-hq/pull/35168#discussion_r1800689494)

```suggestion
                    if not has_non_v2_reference(domain, app):
```

File: corehq/mobile_ucr_v2_update_script.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        if domain in skip_domains:
+            save_in_log(f"Skipped domain: {domain}")
+            continue
+
+        save_in_log(f"Processing domain: {domain} ...")
+        app_ids = list(get_latest_app_ids_and_versions(domain))
+        apps = get_apps_by_id(domain, app_ids)
+        for app in apps:
+            try:
+                # Don't look at app.is_released since the latest version might not be released yet
```
Comment (ajeety4) [inline corehq/mobile_ucr_v2_update_script.py:63]: It would also be good to check before moving further if the app has the reporting modules included which is required step for integrating mobile ucr. It could be more efficient for large apps that do not have any references.

```
if not list(app.get_report_modules()):
  continue

**Review by ajeety4 — CHANGES_REQUESTED** [review]
Requested some changes based on latest findings as explained [here](https://github.com/dimagi/commcare-hq/pull/35168#discussion_r1800689494)

**Comment by ajeety4** [comment]
Hey @zandre-eng ,

Wanted to add one more thing based on earlier script run for finding non-v2 apps. 
The django machine for India is quite small (2 GB) and as a result, this script will cause memory error. 
So we will need to do some slight modifications before running this for India.  The modifications I made are in [this gist](https://gist.github.com/ajeety4/1c08c884cf3854d13decc3ddf81dc34b).  I can offline with you on this.

File: corehq/mobile_ucr_v2_update_script.py (hunk 8 above)
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py]: That sounds like a good idea. Addressed in [25a0298](https://github.com/dimagi/commcare-hq/pull/35158/commits/25a0298903c43d9c7dc0f9e8fe6d95ab7e4e678c).

File: corehq/mobile_ucr_v2_update_script.py (hunk 10 above)
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py:30]: Thanks for flagging this, I have updated the script to also check for case list references in [6dcc719](https://github.com/dimagi/commcare-hq/pull/35158/commits/6dcc7195eb8f0401f068c92531e9c24a3f9ba031).

File: corehq/mobile_ucr_v2_update_script.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    with open(path, 'a') as file:
+        print(json.dumps(data, separators=(',', ':')), file=file)
+
+
+def read_ndjson_file(path):
+    with open(path, 'r') as file:
+        return set(json.loads(line) for line in file.readlines())
+
+
+def has_non_v2_form(domain, app):
```
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py]: Addressed in [6dcc719](https://github.com/dimagi/commcare-hq/pull/35158/commits/6dcc7195eb8f0401f068c92531e9c24a3f9ba031).

File: corehq/mobile_ucr_v2_update_script.py (hunk 11 above)
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py]: Addressed in [6dcc719](https://github.com/dimagi/commcare-hq/pull/35158/commits/6dcc7195eb8f0401f068c92531e9c24a3f9ba031).

File: corehq/mobile_ucr_v2_update_script.py (hunk 12 above)
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py:74]: This is a good idea, and can help prevent needless loop iterations. Addressed in [f86eeef](https://github.com/dimagi/commcare-hq/pull/35158/commits/f86eeef663e85dfa01cd5574a555940403a6438f).

**Comment by zandre-eng** [comment]
> Wanted to add one more thing based on earlier script run for finding non-v2 apps.
The django machine for India is quite small (2 GB) and as a result, this script will cause memory error.
So we will need to do some slight modifications before running this for India. The modifications I made are in [this gist](https://gist.github.com/ajeety4/1c08c884cf3854d13decc3ddf81dc34b). I can offline with you on this.

@ajeety4 Ah this is good to know, thanks for flagging. I have made a few adjustments based on the gist you provided ([241fd96](https://github.com/dimagi/commcare-hq/pull/35158/commits/241fd96ec4c76cbf30babb80687c6c452e6e3034)). Let me know if it looks good, or if there was something else I missed. Thanks!

File: corehq/mobile_ucr_v2_update_script.py (hunk 5 above)
Comment (zandre-eng) [reply corehq/mobile_ucr_v2_update_script.py:74]: We are getting the applications themselves using `get_apps_by_id()` which returns a list of `Application` objects. Since we are only getting the latest release of each app, these retrieved application objects should be the current build of the app.

@mkangia I'm not sure if the above answers your question, but let me know if it doesn't and I'd be happy to offline on this.

Comment (mkangia) [reply corehq/mobile_ucr_v2_update_script.py:74]: > Since we are only getting the latest release of each app, these retrieved application objects should be the current build of the app.

I think there is a misunderstanding of how apps work.

So, an application is actually a running doc and the versions are saved instances of that application at different points(versions). My understanding is that here we want to update the application and not the latest created version/build.

Seems like we are okay here since you are using `get_latest_app_ids_and_versions` which is using app_manager/applications_brief couch view which only keeps applications and not builds since it only saves docs that have [copy_of set as null](https://github.com/dimagi/commcare-hq/blob/73d612ec2bd6dce6572dde770dfbde084495369d/corehq/apps/app_manager/_design/views/applications_brief/map.js#L2)

**Comment by ajeety4** [comment]
> > Wanted to add one more thing based on earlier script run for finding non-v2 apps.
> > The django machine for India is quite small (2 GB) and as a result, this script will cause memory error.
> > So we will need to do some slight modifications before running this for India. The modifications I made are in [this gist](https://gist.github.com/ajeety4/1c08c884cf3854d13decc3ddf81dc34b). I can offline with you on this.
> 
> @ajeety4 Ah this is good to know, thanks for flagging. I have made a few adjustments based on the gist you provided ([241fd96](https://github.com/dimagi/commcare-hq/pull/35158/commits/241fd96ec4c76cbf30babb80687c6c452e6e3034)). Let me know if it looks good, or if there was something else I missed. Thanks!

Hey @zandre-eng , the changes looks good. Thanks for adding this. FYI for the run, I had set the memory limit to 900 MB.
 
One more adjustment I had to make to get the successful run for some of the domains was below, basically fetching one app at at time. Could be done now or while running the script. 

Instead of below, 
```
        apps = get_apps_by_id(domain, app_ids)
        for app in apps:
```
I used this
```
        for app_id in app_ids:
            app = get_apps_by_id(domain, [app_id])[0]
```

**Review by ajeety4 — APPROVED** [review]

**Comment by zandre-eng** [comment]
> FYI for the run, I had set the memory limit to 900 MB.

That's good to know, thanks!

> One more adjustment I had to make to get the successful run for some of the domains was below, basically fetching one app at at time. Could be done now or while running the script.

@ajeety4 Was this specifically for the India environment? Doing it in this manner will incur additional DB hits since we'll know need to retrieve each app one at a time. I think if this is specific to only the India env then I'll only let this adjustment run for them and keep the current functionality for the production env. This will help to minimize the amount of DB calls for envs that have a larger working memory.

**Comment by ajeety4** [comment]
> > FYI for the run, I had set the memory limit to 900 MB.
> 
> That's good to know, thanks!
> 
> > One more adjustment I had to make to get the successful run for some of the domains was below, basically fetching one app at at time. Could be done now or while running the script.
> 
> @ajeety4 Was this specifically for the India environment? Doing it in this manner will incur additional DB hits since we'll know need to retrieve each app one at a time. I think if this is specific to only the India env then I'll only let this adjustment run for them and keep the current functionality for the production env. This will help to minimize the amount of DB calls for envs that have a larger working memory.

Yes specific to the India environment. I agree with keeping the current code for the production environment and doing the adjustment at the run time sounds good.

**Comment by zandre-eng** [comment]
> Yes specific to the India environment. I agree with keeping the current code for the production environment and doing the adjustment at the run time sounds good.

@ajeety4 I decided to be more explicit about it and implement code for both workflows so that it's more clear what needs to change for low memory envs ([0833b6f](https://github.com/dimagi/commcare-hq/pull/35158/commits/0833b6f5df6b69a8402aa6cdced6d78704b26c76)).
